src/Compilers/CSharp/Test/Semantic/Semantics/RefExtensionMethodsTests.cs

```diff
@@ -0,0 +1,1844 @@
+                //         var value1 = obj?.GetValue();        // This should be an error
+                Diagnostic(ErrorCode.ERR_BadUnaryOp, "?").WithArguments("?", "TestType").WithLocation(10, 25));
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_ValueTypes_Allowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref this int p)
+    {
+        System.Console.Write(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        int x = 5;
+        x.PrintValue();
+        Extensions.PrintValue(ref x);
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "55");
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_ValueTypes_Allowed_Reference()
+        {
+            var reference = CreateCompilationWithMscorlibAndSystemCore(@"
+public static class Extensions
+{
+    public static void PrintValue(ref this int p)
+    {
+        System.Console.Write(p);
+    }
+}");
+
+            var code = @"
+public static class Program
+{
+    public static void Main()
+    {
+        int x = 5;
+        x.PrintValue();
+        Extensions.PrintValue(ref x);
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { reference.ToMetadataReference() }, expectedOutput: "55");
+            CompileAndVerify(code, additionalRefs: new[] { reference.EmitToImageReference() }, expectedOutput: "55");
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_ReferenceTypes_NotAllowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref this string p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        string x = ""test"";
+        x.PrintValue();
+        Extensions.PrintValue(ref x);
+    }
+}";
+
+            CreateCompilationWithMscorlibAndSystemCore(code).VerifyDiagnostics(
+                // (4,24): error CS8414: The first parameter of the reference extension method 'PrintValue' must be a value type or a generic type constrained to struct.
+                //     public static void PrintValue(ref this string p)
+                Diagnostic(ErrorCode.ERR_RefExtensionMustBeValueTypeOrConstrainedToOne, "PrintValue").WithArguments("PrintValue").WithLocation(4, 24));
+        }
+
+        [Fact]
+        public void RefExtensionMethodsReceiverTypes_InterfaceTypes_NotAllowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref this System.IComparable p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        System.IComparable x = 5;
+        x.PrintValue();
```

Comment:
>x.PrintValue(); [](start = 8, length = 15)

I would expect an error at this use site, the PrintValue could be declared in a referenced compilation and, in that case, there should be no difference to using the method from metadata.